File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.java
import java.util.Deque;
import java.util.LinkedList;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Deque<Integer> waitingQueue = new LinkedList<>();
        Deque<Integer> bridgeQueue = new LinkedList<>();
        int weights = 0;
        int runningTrucks = 0;
        int time = 0;
        
        for (int truckWeight : truck_weights) {
            waitingQueue.add(truckWeight);
        }
        
        for (int i = 0; i < bridge_length; i++) {
            bridgeQueue.add(null);
        }
        
        while (!waitingQueue.isEmpty() || !bridgeQueue.isEmpty()) {
            time++;
            Integer arrivedTruck = bridgeQueue.pollFirst();
            if (arrivedTruck != null) {
                weights -= arrivedTruck;
                runningTrucks--;
            }
            if (runningTrucks < bridge_length
                && !waitingQueue.isEmpty()
                && waitingQueue.getFirst() + weights <= weight) {
                int truck = waitingQueue.removeFirst();
                bridgeQueue.add(truck);
                weights += truck;
                runningTrucks++;
            } else {
                bridgeQueue.add(null);
            }
            if (runningTrucks == 0 && waitingQueue.isEmpty()) {
                return time;
            }
        }
        
        return time;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.kt
class Solution {
    fun solution(bridge_length: Int, max_weight: Int, truck_weights: IntArray): Int {
        val bridge = ArrayDeque(List(bridge_length) {0})
        var currentWeight = 0
        var passed = 0
        var answer = 0
        var index = 0
        
        while (passed < truck_weights.size) {
            answer++
            
            val passedWeight = bridge.removeFirst()
            
            if (passedWeight > 0) {
                currentWeight -= passedWeight
                passed++
            }
            
            if (index >= truck_weights.size) {
                continue
            }
            
            val weight = truck_weights[index]
            
            if (weight + currentWeight > max_weight) {
                bridge.addLast(0)
            } else {
                bridge.addLast(weight)
                currentWeight += weight
                index++
            }
        }
        
        return answer
    }
}